File: src/main/python/day01.py
import fileinput
import math

# --- Day 1: The Tyranny of the Rocket Equation ---
# --- Part one ---

masses = [x.strip() for x in fileinput.input()]
print(sum([math.floor(int(x) / 3) - 2 for x in masses]))


# --- Part two ---

def calc_fuel(mass):
    def f(x): return math.floor(int(x) / 3) - 2

    act_val = mass
    fuel_sum = 0
    while True:
        act_val = f(act_val)
        if act_val < 0:
            break

        fuel_sum += act_val
    return fuel_sum


assert calc_fuel(100756) == 50346

print(sum([calc_fuel(mass) for mass in masses]))

File: src/main/R/day01.R

## --- Day 1: The Tyranny of the Rocket Equation ---

## --- Part one ---

library(data.table)
library(testthat)

input <- fread("../resources/day01input.txt", header = FALSE, sep = "\n")$V1
sum(floor(input / 3) - 2)

## 3391707


## --- Part two ---

calc_fuel <- function(x) {
  f <- function(y) { floor(y / 3) - 2 }

  act_val <- x
  fuel_sum <- 0
  while (TRUE) {
    act_val <- f(act_val)
    if (act_val < 0) {
      break
    }
    
    fuel_sum <- fuel_sum + act_val
  }
  
  return(fuel_sum)
}

expect_equal(calc_fuel(100756), 50346)

sum(sapply(input, calc_fuel))

## 5084676

